File: fabric/src/main/java/dev/obscuria/fragmentum/fabric/service/FabricClientService.java
package dev.obscuria.fragmentum.fabric.service;

import dev.obscuria.fragmentum.fabric.registry.FabricClientRegistrar;
import dev.obscuria.fragmentum.service.ClientService;
import dev.obscuria.fragmentum.v2.api.client.ClientRegistrar;

public final class FabricClientService implements ClientService {

    public static final FabricClientService SHARED = new FabricClientService();

    @Override public ClientRegistrar registrar(String modId) {
        return new FabricClientRegistrar(modId);
    }

    private FabricClientService() {}
}

File: fabric/src/main/kotlin/dev/obscuria/fragmentum/fabric/service/FabricClientService.kt
package dev.obscuria.fragmentum.fabric.service

import dev.obscuria.fragmentum.client.ClientRegistrar
import dev.obscuria.fragmentum.service.ClientService
import dev.obscuria.fragmentum.client.TooltipComponentRegistry
import dev.obscuria.fragmentum.fabric.impl.FabricClientRegistrar
import net.minecraft.client.gui.screens.inventory.tooltip.ClientTooltipComponent
import net.minecraft.world.inventory.tooltip.TooltipComponent
import java.util.function.Function

internal object FabricClientService : ClientService {

    override fun registrar(modId: String): ClientRegistrar {
        return FabricClientRegistrar(modId)
    }

    override fun <T : TooltipComponent> registerTooltipComponent(
        type: Class<T>,
        factory: Function<T, ClientTooltipComponent>
    ) {
        TooltipComponentRegistry.register(type, factory)
    }
}